crates/router/src/core/payment_methods/cards.rs

```
@@ -216,6 +225,208 @@ pub async fn get_or_insert_payment_method(
+    } else {
+        let current_timestamp = common_utils::date_time::now();
+        let session_expiry = payment_method
+            .created_at
+            .saturating_add(time::Duration::seconds(consts::DEFAULT_SESSION_EXPIRY));
+
+        fp_utils::when(current_timestamp > session_expiry, || {
+            Err::<(), errors::ApiErrorResponse>(errors::ApiErrorResponse::ClientSecretExpired)
+        })?;
+
+        Ok(payment_method)
+    }
+}
+
+#[instrument(skip_all)]
+pub async fn add_payment_method_data(
+    state: routes::AppState,
+    req: api::PaymentMethodCreate,
+    merchant_account: domain::MerchantAccount,
+    key_store: domain::MerchantKeyStore,
+    pm_id: String,
+) -> errors::RouterResponse<api::PaymentMethodResponse> {
+    let db = &*state.store;
+    let customer_id = req.customer_id.clone().get_required_value("customer_id")?;
+
+    let pmd = req
+        .payment_method_data
+        .clone()
+        .get_required_value("payment_method_data")?;
+    let _payment_method = req.payment_method.get_required_value("payment_method")?;
+    let client_secret = req
+        .client_secret
+        .clone()
+        .get_required_value("client_secret")?;
+
+    let payment_method = authenticate_pm_client_secret(&client_secret, db, pm_id.as_str()).await?;
+
+    match pmd {
+        api_models::payment_methods::PaymentMethodCreateData::Card(card) => {
+            let resp =
+                add_card_to_locker(&state, req.clone(), &card, &customer_id, &merchant_account)
+                    .await
+                    .change_context(errors::ApiErrorResponse::InternalServerError);
+
+            match resp {
+                Ok((mut pm_resp, duplication_check)) => {
+                    if duplication_check.is_some() {
+                        let pm_update = storage::PaymentMethodUpdate::StatusUpdate {
+                            status: Some(enums::PaymentMethodStatus::Inactive),
+                        };
```

@vspecky we can return an error in this case right? we need not make the payment_method as inactive and send some other payment method.